flang/include/flang/Lower/IntrinsicCall.h

```diff
@@ -37,6 +29,35 @@ fir::ExtendedValue genIntrinsicCall(FirOpBuilder &, mlir::Location,
                                     llvm::Optional<mlir::Type> resultType,
                                     llvm::ArrayRef<fir::ExtendedValue> args);
 
+/// Enum specifying how intrinsic argument evaluate::Expr should be
+/// lowered to fir::ExtendedValue to be passed to genIntrinsicCall.
+enum class LowerIntrinsicArgAs {
+  /// Lower argument to a value. Mainly intended for scalar arguments.
+  Value,
+  /// Lower argument to an address. Only valid when the argument properties are
+  /// fully
+  /// defined (e.g. allocatable is allocated...).
+  Addr,
+  /// Lower argument without assuming that the argument is fully defined.
+  /// It can be used on unallocated allocatable, disassociated pointer,
+  /// or absent optional.
+  Inquired
```

Comment:
`Nullable`? Inquired struck me as odd at first, but I can get used to it. It's less suggestive of an implementation artifact, so that's a `+`.

Reply:
Yes, `Inquired` is a bit weird. It is really meant to be used for all the "inquired" argument of inquiry intrinsic, I have added "This is meant for inquiry intrinsic arguments" to clarify the origin of the name.
The logic/naming here might anyway be refined when adding more intrinsic.

Reply:
Ah. OK, we may indeed not want to overload this value to mean different things then.